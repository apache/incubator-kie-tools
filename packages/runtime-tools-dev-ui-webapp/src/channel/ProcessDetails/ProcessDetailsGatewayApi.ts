/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ProcessDetailsQueries } from "./ProcessDetailsQueries";
import {
  ProcessInstance,
  SvgSuccessResponse,
  SvgErrorResponse,
  TriggerableNode,
  NodeInstance,
} from "@kogito-apps/management-console-shared/dist/types";

export interface OnOpenProcessInstanceDetailsListener {
  onOpen(id: string): void;
}

export interface ProcessDetailsUnSubscribeHandler {
  unSubscribe: () => void;
}

export interface ProcessDetailsState {
  id: string;
}

export interface ProcessDetailsGatewayApi {
  processDetailsState: any;
  getProcessDiagram: (data: ProcessInstance) => Promise<SvgSuccessResponse | SvgErrorResponse>;
  handleProcessAbort: (processInstance: ProcessInstance) => Promise<void>;
  getTriggerableNodes(processInstance: ProcessInstance): Promise<TriggerableNode[]>;
  handleNodeTrigger(processInstance: ProcessInstance, node: any): Promise<void>;
  handleProcessVariableUpdate: (
    processInstance: ProcessInstance,
    updateJson: Record<string, unknown>
  ) => Promise<Record<string, unknown>>;
  processDetailsQuery(id: string): Promise<ProcessInstance>;
  openProcessInstanceDetails(id: string): Promise<void>;
  onOpenProcessInstanceDetailsListener: (
    listener: OnOpenProcessInstanceDetailsListener
  ) => ProcessDetailsUnSubscribeHandler;
  handleProcessRetry: (processInstance: ProcessInstance) => Promise<void>;
  handleNodeInstanceCancel: (processInstance: ProcessInstance, node: NodeInstance) => Promise<void>;
  handleProcessSkip: (processInstance: ProcessInstance) => Promise<void>;
  handleNodeInstanceRetrigger(processInstance: ProcessInstance, node: NodeInstance): Promise<void>;
}

export class ProcessDetailsGatewayApiImpl implements ProcessDetailsGatewayApi {
  private readonly queries: ProcessDetailsQueries;
  private _ProcessDetailsState: ProcessDetailsState;
  private readonly listeners: OnOpenProcessInstanceDetailsListener[] = [];

  constructor(queries: ProcessDetailsQueries) {
    this.queries = queries;
    this._ProcessDetailsState = { id: "" };
  }

  get processDetailsState(): ProcessDetailsState {
    return this._ProcessDetailsState;
  }

  getProcessDiagram = async (data: ProcessInstance): Promise<SvgSuccessResponse | SvgErrorResponse> => {
    const result = await this.queries.getSVG(data);
    return Promise.resolve(result);
  };

  handleProcessAbort = (processInstance: ProcessInstance): Promise<void> => {
    return this.queries.handleProcessAbort(processInstance);
  };

  getTriggerableNodes(processInstance: ProcessInstance): Promise<TriggerableNode[]> {
    return this.queries.getTriggerableNodes(processInstance);
  }

  handleNodeTrigger(processInstance: ProcessInstance, node: TriggerableNode): Promise<void> {
    return this.queries.handleNodeTrigger(processInstance, node);
  }

  handleProcessVariableUpdate = (processInstance: ProcessInstance, updatedJson: Record<string, unknown>) => {
    return this.queries.handleProcessVariableUpdate(processInstance, updatedJson);
  };

  processDetailsQuery(id: string): Promise<ProcessInstance> {
    return new Promise<any>((resolve, reject) => {
      this.queries
        .getProcessDetails(id)
        .then((value) => {
          resolve(value);
        })
        .catch((reason) => {
          reject(reason);
        });
    });
  }

  openProcessInstanceDetails(id: string): Promise<void> {
    this._ProcessDetailsState = { id: id };
    this.listeners.forEach((listener) => listener.onOpen(id));
    return Promise.resolve();
  }

  onOpenProcessInstanceDetailsListener(
    listener: OnOpenProcessInstanceDetailsListener
  ): ProcessDetailsUnSubscribeHandler {
    this.listeners.push(listener);

    const unSubscribe = () => {
      const index = this.listeners.indexOf(listener);
      if (index > -1) {
        this.listeners.splice(index, 1);
      }
    };

    return {
      unSubscribe,
    };
  }

  handleProcessRetry(processInstance: ProcessInstance): Promise<void> {
    return this.queries.handleProcessRetry(processInstance);
  }

  handleNodeInstanceCancel(processInstance: ProcessInstance, node: NodeInstance): Promise<void> {
    return this.queries.handleNodeInstanceCancel(processInstance, node);
  }

  handleProcessSkip(processInstance: ProcessInstance): Promise<void> {
    return this.queries.handleProcessSkip(processInstance);
  }

  handleNodeInstanceRetrigger(processInstance: ProcessInstance, node: NodeInstance): Promise<void> {
    return this.queries.handleNodeInstanceRetrigger(processInstance, node);
  }
}
