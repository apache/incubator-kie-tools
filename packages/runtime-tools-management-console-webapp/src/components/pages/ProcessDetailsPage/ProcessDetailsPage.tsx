/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import * as React from "react";
import { useState, useEffect } from "react";
import { Card } from "@patternfly/react-core/dist/js/components/Card";
import { PageSection } from "@patternfly/react-core/dist/js/components/Page";
import { Bullseye } from "@patternfly/react-core/dist/js/layouts/Bullseye";
import { RouteComponentProps } from "react-router-dom";
import ProcessDetailsContainer from "../../containers/ProcessDetailsContainer/ProcessDetailsContainer";
import { StaticContext, useHistory } from "react-router";
import * as H from "history";
import "../../styles.css";
import { OUIAProps, ouiaPageTypeAndObjectId } from "@kie-tools/runtime-tools-components/dist/ouiaTools";
import { ProcessInstance } from "@kie-tools/runtime-tools-process-gateway-api/dist/types";
import { ServerErrors } from "@kie-tools/runtime-tools-components/dist/components/ServerErrors";
import { KogitoSpinner } from "@kie-tools/runtime-tools-components/dist/components/KogitoSpinner";
import { PageSectionHeader } from "@kie-tools/runtime-tools-components/dist/components/PageSectionHeader";
import {
  useProcessDetailsGatewayApi,
  ProcessDetailsGatewayApi,
} from "@kie-tools/runtime-tools-process-webapp-components/dist/ProcessDetails";

interface MatchProps {
  instanceID: string;
}

const ProcessDetailsPage: React.FC<RouteComponentProps<MatchProps, StaticContext, H.LocationState> & OUIAProps> = ({
  ...props
}) => {
  const processId = props.match.params.instanceID;
  useEffect(() => {
    return ouiaPageTypeAndObjectId("process-instances", processId);
  });

  const gatewayApi: ProcessDetailsGatewayApi = useProcessDetailsGatewayApi();

  const history = useHistory();
  const [processInstance, setProcessInstance] = useState<ProcessInstance>({} as ProcessInstance);
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [error, setError] = useState<string>("");
  let currentPage = JSON.parse(window.localStorage.getItem("state"));

  useEffect(() => {
    window.onpopstate = () => {
      props.history.push({ state: Object.assign({}, props.location.state) });
    };
  });

  async function fetchDetails() {
    let response: ProcessInstance = {} as ProcessInstance;
    try {
      setIsLoading(true);
      response = await gatewayApi.processDetailsQuery(processId);
      setProcessInstance(response);
    } catch (error) {
      setError(error);
    } finally {
      setIsLoading(false);
      /* istanbul ignore else */
      if (error.length === 0 && Object.keys(response).length === 0) {
        let prevPath;
        /* istanbul ignore else */
        if (currentPage) {
          currentPage = Object.assign({}, currentPage, props.location.state);
          const tempPath = currentPage.prev.split("/");
          prevPath = tempPath.filter((item) => item);
        }
        history.push({
          pathname: "/NoData",
          state: {
            prev: currentPage ? currentPage.prev : "/ProcessInstances",
            title: "Process not found",
            description: `Process instance with the id ${processId} not found`,
            buttonText: currentPage
              ? `Go to ${prevPath[0]
                  .replace(/([A-Z])/g, " $1")
                  .trim()
                  .toLowerCase()}`
              : "Go to process instances",
            rememberedData: Object.assign({}, props.location.state),
          },
        });
      }
    }
  }

  useEffect(() => {
    /* istanbul ignore else */
    if (processId) {
      fetchDetails();
    }
  }, [processId]);

  const processName = processInstance ? processInstance.processName : "";

  const renderItems = () => {
    if (!isLoading) {
      return (
        <>
          {processInstance && Object.keys(processInstance).length > 0 && !error ? (
            <ProcessDetailsContainer processInstance={processInstance} />
          ) : (
            <>
              {error.length > 0 && (
                <Card className="kogito-management-console__card-size">
                  <Bullseye>
                    <ServerErrors error={error} variant="large" />
                  </Bullseye>
                </Card>
              )}
            </>
          )}
        </>
      );
    } else {
      return (
        <Card>
          <KogitoSpinner spinnerText="Loading process details..." />
        </Card>
      );
    }
  };

  return (
    <React.Fragment>
      <PageSectionHeader
        titleText="Process Details"
        breadcrumbText={["Home", "Processes", processName]}
        breadcrumbPath={[
          "/",
          {
            pathname: "/ProcessInstances",
            state: Object.assign({}, props.location.state),
          },
        ]}
      />
      <PageSection>{renderItems()}</PageSection>
    </React.Fragment>
  );
};

export default ProcessDetailsPage;
