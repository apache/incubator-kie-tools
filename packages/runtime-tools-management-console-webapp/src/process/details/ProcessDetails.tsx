/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import * as React from "react";
import { useState, useEffect, useCallback } from "react";
import { Card } from "@patternfly/react-core/dist/js/components/Card";
import { Bullseye } from "@patternfly/react-core/dist/js/layouts/Bullseye";
import { ProcessInstance } from "@kie-tools/runtime-tools-process-gateway-api/dist/types";
import { ServerErrors } from "@kie-tools/runtime-tools-components/dist/components/ServerErrors";
import { KogitoSpinner } from "@kie-tools/runtime-tools-components/dist/components/KogitoSpinner";
import {
  EmptyState,
  EmptyStateBody,
  EmptyStateIcon,
  EmptyStateVariant,
  EmptyStateHeader,
  EmptyStateFooter,
} from "@patternfly/react-core/dist/js/components/EmptyState";
import { SearchIcon } from "@patternfly/react-icons/dist/js/icons/search-icon";
import { Button } from "@patternfly/react-core/dist/js/components/Button";
import { useCancelableEffect } from "@kie-tools-core/react-hooks/dist/useCancelableEffect";
import { EmbeddedProcessDetails } from "@kie-tools/runtime-tools-process-enveloped-components/dist/processDetails";
import { useProcessDetailsChannelApi } from "@kie-tools/runtime-tools-process-webapp-components/dist/ProcessDetails";
import { useRuntimeSpecificRoutes } from "../../runtime/RuntimeContext";
import { useNavigate } from "react-router-dom";

interface Props {
  processInstanceId: string;
  onReturnToProcessList: () => void;
}

export const ProcessDetails: React.FC<Props> = ({ processInstanceId, onReturnToProcessList }) => {
  const channelApi = useProcessDetailsChannelApi();
  const runtimeRoutes = useRuntimeSpecificRoutes();
  const navigate = useNavigate();
  const [processInstance, setProcessInstance] = useState<ProcessInstance>();
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [error, setError] = useState<string>();

  useCancelableEffect(
    useCallback(
      ({ canceled }) => {
        if (canceled.get()) {
          return;
        }
        setIsLoading(true);
        channelApi
          .processDetails__getProcessDetails(processInstanceId)
          .then((response) => {
            if (canceled.get()) {
              return;
            }
            setProcessInstance(response);
          })
          .catch((error) => {
            setError(error);
          })
          .finally(() => {
            setIsLoading(false);
          });
      },
      [channelApi, processInstanceId]
    )
  );

  useEffect(() => {
    const unsubscriber = channelApi.processDetails__onOpenProcessInstanceDetailsListener({
      onOpen(id: string) {
        navigate(runtimeRoutes.processDetails(id));
      },
    });

    return () => {
      unsubscriber.then((unsubscribeHandler) => unsubscribeHandler.unSubscribe());
    };
  }, [channelApi, navigate, runtimeRoutes]);

  // Loading State
  if (isLoading) {
    return (
      <Card>
        <KogitoSpinner spinnerText="Loading process details..." />
      </Card>
    );
  }

  // Error State
  if (error) {
    return <ServerErrors error={error} variant="large" onGoBack={() => navigate(runtimeRoutes.processes())} />;
  }

  // Process Instance Details
  if (processInstance) {
    return (
      <EmbeddedProcessDetails
        channelApi={channelApi}
        targetOrigin={window.location.origin}
        processInstance={processInstance}
        singularProcessLabel={"process"}
        pluralProcessLabel={"processes"}
      />
    );
  }

  return (
    <Bullseye>
      <EmptyState variant={EmptyStateVariant.full}>
        <EmptyStateHeader
          titleText="Process instance not found"
          icon={<EmptyStateIcon icon={SearchIcon} />}
          headingLevel="h1"
        />
        <EmptyStateBody>{`Process instance with the id ${processInstanceId} not found`}</EmptyStateBody>
        <EmptyStateFooter>
          <Button variant="primary" onClick={onReturnToProcessList} data-testid="redirect-button">
            Back to Process Instances
          </Button>
        </EmptyStateFooter>
      </EmptyState>
    </Bullseye>
  );
};
