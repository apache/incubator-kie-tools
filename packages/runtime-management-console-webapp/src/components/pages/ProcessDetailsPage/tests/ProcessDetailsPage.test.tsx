/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React from "react";
import * as H from "history";
import { mount } from "enzyme";
import {
  MilestoneStatus,
  ProcessInstance,
  ProcessInstanceState,
} from "@kie-tools/runtime-tools-process-gateway-api/dist/types";
import ProcessDetailsPage from "../ProcessDetailsPage";
import { BrowserRouter } from "react-router-dom";
import * as ProcessDetailsContext from "../../../../channel/ProcessDetails/ProcessDetailsContext";
import { ProcessDetailsGatewayApi } from "../../../../channel/ProcessDetails/ProcessDetailsGatewayApi";
import { act } from "react-dom/test-utils";
import wait from "waait";
jest.mock("../../../containers/ProcessDetailsContainer/ProcessDetailsContainer");
describe("WebApp - ProcessDetailsPage tests", () => {
  const props = {
    match: {
      params: {
        instanceID: "8035b580-6ae4-4aa8-9ec0-e18e19809e0b",
      },
      url: "",
      isExact: false,
      path: "",
    },
    location: H.createLocation(""),
    history: H.createBrowserHistory(),
  };

  const data: ProcessInstance = {
    id: "8035b580-6ae4-4aa8-9ec0-e18e19809e0b",
    processId: "hotelBooking",
    processName: "HotelBooking",
    businessKey: "T1234HotelBooking01",
    parentProcessInstanceId: "e4448857-fa0c-403b-ad69-f0a353458b9d",
    parentProcessInstance: null,
    roles: [],
    variables:
      '{"trip":{"begin":"2019-10-22T22:00:00Z[UTC]","city":"Bangalore","country":"India","end":"2019-10-30T22:00:00Z[UTC]","visaRequired":false},"hotel":{"address":{"city":"Bangalore","country":"India","street":"street","zipCode":"12345"},"bookingNumber":"XX-012345","name":"Perfect hotel","phone":"09876543"},"traveller":{"address":{"city":"Bangalore","country":"US","street":"Bangalore","zipCode":"560093"},"email":"ajaganat@redhat.com","firstName":"Ajay","lastName":"Jaganathan","nationality":"US"}}',
    state: ProcessInstanceState.Completed,
    start: new Date("2019-10-22T03:40:44.089Z"),
    lastUpdate: new Date("Thu, 22 Apr 2021 14:53:04 GMT"),
    end: new Date("2019-10-22T05:40:44.089Z"),
    addons: ["process-management"],
    endpoint: "http://localhost:4000",
    serviceUrl: "http://localhost:4000",
    error: {
      nodeDefinitionId: "a1e139d5-4e77-48c9-84ae-34578e904e6b",
      message: "some thing went wrong",
      __typename: "ProcessInstanceError",
    },
    childProcessInstances: [],
    nodes: [
      {
        id: "27107f38-d888-4edf-9a4f-11b9e6d751b6",
        nodeId: "1",
        name: "End Event 1",
        enter: new Date("2019-10-22T03:37:30.798Z"),
        exit: new Date("2019-10-22T03:37:30.798Z"),
        type: "EndNode",
        definitionId: "EndEvent_1",
        __typename: "NodeInstance",
      },
      {
        id: "41b3f49e-beb3-4b5f-8130-efd28f82b971",
        nodeId: "2",
        name: "Book hotel",
        enter: new Date("2019-10-22T03:37:30.795Z"),
        exit: new Date("2019-10-22T03:37:30.798Z"),
        type: "WorkItemNode",
        definitionId: "ServiceTask_1",
        __typename: "NodeInstance",
      },
      {
        id: "4165a571-2c79-4fd0-921e-c6d5e7851b67",
        nodeId: "2",
        name: "StartProcess",
        enter: new Date("2019-10-22T03:37:30.793Z"),
        exit: new Date("2019-10-22T03:37:30.795Z"),
        type: "StartNode",
        definitionId: "StartEvent_1",
        __typename: "NodeInstance",
      },
    ],
    milestones: [
      {
        id: "27107f38-d888-4edf-9a4f-11b9e6d75i86",
        name: "Manager decision",
        status: MilestoneStatus.Completed,
        __typename: "Milestone",
      },
      {
        id: "27107f38-d888-4edf-9a4f-11b9e6d75m36",
        name: "Milestone 1: Order placed",
        status: MilestoneStatus.Active,
        __typename: "Milestone",
      },
      {
        id: "27107f38-d888-4edf-9a4f-11b9e6d75m66",
        name: "Milestone 2: Order shipped",
        status: MilestoneStatus.Available,
        __typename: "Milestone",
      },
    ],
  };

  const getProcessDetails = jest.fn();
  const MockProcessDetailsGatewayApi = jest.fn<ProcessDetailsGatewayApi, []>(() => ({
    getProcessDiagram: jest.fn(),
    handleProcessAbort: jest.fn(),
    cancelJob: jest.fn(),
    rescheduleJob: jest.fn(),
    jobsQuery: jest.fn(),
    handleProcessVariableUpdate: jest.fn(),
    getTriggerableNodes: jest.fn(),
    handleNodeTrigger: jest.fn(),
    openProcessInstanceDetails: jest.fn(),
    onOpenProcessInstanceDetailsListener: jest.fn(),
    processDetailsQuery: getProcessDetails,
    processDetailsState: undefined,
    handleProcessRetry: jest.fn(),
    handleNodeInstanceCancel: jest.fn(),
    handleProcessSkip: jest.fn(),
    handleNodeInstanceRetrigger: jest.fn(),
  }));
  let gatewayApi: ProcessDetailsGatewayApi;

  jest.spyOn(ProcessDetailsContext, "useProcessDetailsGatewayApi").mockImplementation(() => gatewayApi);

  beforeEach(() => {
    jest.clearAllMocks();
    gatewayApi = new MockProcessDetailsGatewayApi();
  });
  it("Snapshot test with default values", async () => {
    (getProcessDetails as jest.Mock).mockReturnValue(data);
    let wrapper;
    await act(async () => {
      wrapper = mount(
        <BrowserRouter>
          <ProcessDetailsPage {...props} />
        </BrowserRouter>
      );
      await wait(0);
      wrapper = wrapper.update().find("ProcessDetailsPage");
    });
    expect(wrapper).toMatchSnapshot();
  });
});
