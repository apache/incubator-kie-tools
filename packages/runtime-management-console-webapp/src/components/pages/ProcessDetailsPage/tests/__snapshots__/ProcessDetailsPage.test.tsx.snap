// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebApp - ProcessDetailsPage tests Snapshot test with default values 1`] = `
<ProcessDetailsPage
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "",
      "pathname": "/",
      "search": "",
      "state": undefined,
    }
  }
  match={
    Object {
      "isExact": false,
      "params": Object {
        "instanceID": "8035b580-6ae4-4aa8-9ec0-e18e19809e0b",
      },
      "path": "",
      "url": "",
    }
  }
>
  <PageSectionHeader
    breadcrumbPath={
      Array [
        "/",
        Object {
          "pathname": "/ProcessInstances",
          "state": Object {},
        },
      ]
    }
    breadcrumbText={
      Array [
        "Home",
        "Processes",
        "HotelBooking",
      ]
    }
    titleText="Process Details"
  >
    <PageSection
      data-ouia-component-type="page-section-header"
      data-ouia-safe={true}
      variant="light"
    >
      <section
        className="pf-c-page__main-section pf-m-light"
        data-ouia-component-type="page-section-header"
        data-ouia-safe={true}
      >
        <Breadcrumb>
          <nav
            aria-label="Breadcrumb"
            className="pf-c-breadcrumb"
            data-ouia-component-id="OUIA-Generated-Breadcrumb-1"
            data-ouia-component-type="PF4/Breadcrumb"
            data-ouia-safe={true}
          >
            <ol
              className="pf-c-breadcrumb__list"
            >
              <BreadcrumbItem
                key=".$0"
                showDivider={false}
              >
                <li
                  className="pf-c-breadcrumb__item"
                >
                  <Link
                    to="/"
                  >
                    <LinkAnchor
                      href="/"
                      navigate={[Function]}
                    >
                      <a
                        href="/"
                        onClick={[Function]}
                      >
                        Home
                      </a>
                    </LinkAnchor>
                  </Link>
                </li>
              </BreadcrumbItem>
              <BreadcrumbItem
                key=".$1"
                showDivider={true}
              >
                <li
                  className="pf-c-breadcrumb__item"
                >
                  <span
                    className="pf-c-breadcrumb__item-divider"
                  >
                    <AngleRightIcon
                      color="currentColor"
                      noVerticalAlign={false}
                      size="sm"
                    >
                      <svg
                        aria-hidden={true}
                        aria-labelledby={null}
                        fill="currentColor"
                        height="1em"
                        role="img"
                        style={
                          Object {
                            "verticalAlign": "-0.125em",
                          }
                        }
                        viewBox="0 0 256 512"
                        width="1em"
                      >
                        <path
                          d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                        />
                      </svg>
                    </AngleRightIcon>
                  </span>
                  <Link
                    to={
                      Object {
                        "pathname": "/ProcessInstances",
                        "state": Object {},
                      }
                    }
                  >
                    <LinkAnchor
                      href="/ProcessInstances"
                      navigate={[Function]}
                    >
                      <a
                        href="/ProcessInstances"
                        onClick={[Function]}
                      >
                        Processes
                      </a>
                    </LinkAnchor>
                  </Link>
                </li>
              </BreadcrumbItem>
              <BreadcrumbItem
                isActive={true}
                key=".$2"
                showDivider={true}
              >
                <li
                  className="pf-c-breadcrumb__item"
                >
                  <span
                    className="pf-c-breadcrumb__item-divider"
                  >
                    <AngleRightIcon
                      color="currentColor"
                      noVerticalAlign={false}
                      size="sm"
                    >
                      <svg
                        aria-hidden={true}
                        aria-labelledby={null}
                        fill="currentColor"
                        height="1em"
                        role="img"
                        style={
                          Object {
                            "verticalAlign": "-0.125em",
                          }
                        }
                        viewBox="0 0 256 512"
                        width="1em"
                      >
                        <path
                          d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                        />
                      </svg>
                    </AngleRightIcon>
                  </span>
                  HotelBooking
                </li>
              </BreadcrumbItem>
            </ol>
          </nav>
        </Breadcrumb>
        <PageTitle
          title="Process Details"
        >
          <Flex
            data-ouia-component-type="page-title"
            data-ouia-safe={true}
          >
            <div
              className="pf-l-flex"
              data-ouia-component-type="page-title"
              data-ouia-safe={true}
            >
              <FlexItem
                spacer={
                  Object {
                    "default": "spacerSm",
                  }
                }
              >
                <div
                  className="pf-m-spacer-sm"
                >
                  <Title
                    headingLevel="h1"
                    size="4xl"
                  >
                    <h1
                      className="pf-c-title pf-m-4xl"
                      data-ouia-component-id="OUIA-Generated-Title-1"
                      data-ouia-component-type="PF4/Title"
                      data-ouia-safe={true}
                    >
                      Process Details
                    </h1>
                  </Title>
                </div>
              </FlexItem>
            </div>
          </Flex>
        </PageTitle>
      </section>
    </PageSection>
  </PageSectionHeader>
  <PageSection>
    <section
      className="pf-c-page__main-section"
    >
      <MockedProcessDetailsContainer
        processInstance={
          Object {
            "addons": Array [
              "process-management",
            ],
            "businessKey": "T1234HotelBooking01",
            "childProcessInstances": Array [],
            "end": 2019-10-22T05:40:44.089Z,
            "endpoint": "http://localhost:4000",
            "error": Object {
              "__typename": "ProcessInstanceError",
              "message": "some thing went wrong",
              "nodeDefinitionId": "a1e139d5-4e77-48c9-84ae-34578e904e6b",
            },
            "id": "8035b580-6ae4-4aa8-9ec0-e18e19809e0b",
            "lastUpdate": 2021-04-22T14:53:04.000Z,
            "milestones": Array [
              Object {
                "__typename": "Milestone",
                "id": "27107f38-d888-4edf-9a4f-11b9e6d75i86",
                "name": "Manager decision",
                "status": "COMPLETED",
              },
              Object {
                "__typename": "Milestone",
                "id": "27107f38-d888-4edf-9a4f-11b9e6d75m36",
                "name": "Milestone 1: Order placed",
                "status": "ACTIVE",
              },
              Object {
                "__typename": "Milestone",
                "id": "27107f38-d888-4edf-9a4f-11b9e6d75m66",
                "name": "Milestone 2: Order shipped",
                "status": "AVAILABLE",
              },
            ],
            "nodes": Array [
              Object {
                "__typename": "NodeInstance",
                "definitionId": "EndEvent_1",
                "enter": 2019-10-22T03:37:30.798Z,
                "exit": 2019-10-22T03:37:30.798Z,
                "id": "27107f38-d888-4edf-9a4f-11b9e6d751b6",
                "name": "End Event 1",
                "nodeId": "1",
                "type": "EndNode",
              },
              Object {
                "__typename": "NodeInstance",
                "definitionId": "ServiceTask_1",
                "enter": 2019-10-22T03:37:30.795Z,
                "exit": 2019-10-22T03:37:30.798Z,
                "id": "41b3f49e-beb3-4b5f-8130-efd28f82b971",
                "name": "Book hotel",
                "nodeId": "2",
                "type": "WorkItemNode",
              },
              Object {
                "__typename": "NodeInstance",
                "definitionId": "StartEvent_1",
                "enter": 2019-10-22T03:37:30.793Z,
                "exit": 2019-10-22T03:37:30.795Z,
                "id": "4165a571-2c79-4fd0-921e-c6d5e7851b67",
                "name": "StartProcess",
                "nodeId": "2",
                "type": "StartNode",
              },
            ],
            "parentProcessInstance": null,
            "parentProcessInstanceId": "e4448857-fa0c-403b-ad69-f0a353458b9d",
            "processId": "hotelBooking",
            "processName": "HotelBooking",
            "roles": Array [],
            "serviceUrl": "http://localhost:4000",
            "start": 2019-10-22T03:40:44.089Z,
            "state": "COMPLETED",
            "variables": "{\\"trip\\":{\\"begin\\":\\"2019-10-22T22:00:00Z[UTC]\\",\\"city\\":\\"Bangalore\\",\\"country\\":\\"India\\",\\"end\\":\\"2019-10-30T22:00:00Z[UTC]\\",\\"visaRequired\\":false},\\"hotel\\":{\\"address\\":{\\"city\\":\\"Bangalore\\",\\"country\\":\\"India\\",\\"street\\":\\"street\\",\\"zipCode\\":\\"12345\\"},\\"bookingNumber\\":\\"XX-012345\\",\\"name\\":\\"Perfect hotel\\",\\"phone\\":\\"09876543\\"},\\"traveller\\":{\\"address\\":{\\"city\\":\\"Bangalore\\",\\"country\\":\\"US\\",\\"street\\":\\"Bangalore\\",\\"zipCode\\":\\"560093\\"},\\"email\\":\\"ajaganat@redhat.com\\",\\"firstName\\":\\"Ajay\\",\\"lastName\\":\\"Jaganathan\\",\\"nationality\\":\\"US\\"}}",
          }
        }
      />
    </section>
  </PageSection>
</ProcessDetailsPage>
`;
